adam-core/src/main/scala/org/bdgenomics/adam/rdd/ShuffleRegionJoin.scala

```diff
@@ -38,488 +35,468 @@ import scala.reflect.ClassTag
-  protected val bins = sc.broadcast(GenomeBins(partitionSize, seqLengths))
+  protected def advanceCache(cache: SetTheoryCache[U, RT, RU],
+                             right: BufferedIterator[(ReferenceRegion, U)],
+                             until: ReferenceRegion)
+  protected def pruneCache(cache: SetTheoryCache[U, RT, RU],
+                           to: ReferenceRegion)
+  protected def postProcessHits(iter: Iterable[U],
+                                currentLeft: T): Iterable[(RT, RU)]
+  protected def finalizeHits(cache: SetTheoryCache[U, RT, RU],
+                             right: BufferedIterator[(ReferenceRegion, U)]): Iterable[(RT, RU)]
+  protected def emptyFn(left: Iterator[(ReferenceRegion, T)],
+                        right: Iterator[(ReferenceRegion, U)]): Iterator[(RT, RU)]
+
+  protected val leftRdd: RDD[(ReferenceRegion, T)]
+  protected val rightRdd: RDD[(ReferenceRegion, U)]
 
   /**
-   * Performs a region join between two RDDs (shuffle join).
-   *
-   * This implementation is shuffle-based, so does not require collecting one side into memory
-   * like BroadcastRegionJoin.  It basically performs a global sort of each RDD by genome position
-   * and then does a sort-merge join, similar to the chromsweep implementation in bedtools.  More
-   * specifically, it first defines a set of bins across the genome, then assigns each object in the
-   * RDDs to each bin that they overlap (replicating if necessary), performs the shuffle, and sorts
-   * the object in each bin.  Finally, each bin independently performs a chromsweep sort-merge join.
+   * Performs a region join between two RDDs (shuffle join). All data should be pre-shuffled and
+   * copartitioned.
    *
-   * @param leftRDD The 'left' side of the join
-   * @param rightRDD The 'right' side of the join
-   * @return An RDD of pairs (x, y), where x is from leftRDD, y is from rightRDD, and the region
+   * @return An RDD of joins (x, y), where x is from leftRDD, y is from rightRDD, and the region
    *         corresponding to x overlaps the region corresponding to y.
    */
-  def partitionAndJoin(
-    leftRDD: RDD[(ReferenceRegion, T)],
-    rightRDD: RDD[(ReferenceRegion, U)]): RDD[(RT, RU)] = {
-
-    // Key each RDD element to its corresponding bin
-    // Elements may be replicated if they overlap multiple bins
-    val keyedLeft: RDD[((ReferenceRegion, Int), T)] =
-      leftRDD.flatMap(kv => {
-        val (region, x) = kv
-        val lo = bins.value.getStartBin(region)
-        val hi = bins.value.getEndBin(region)
-        (lo to hi).map(i => ((region, i), x))
-      })
-    val keyedRight: RDD[((ReferenceRegion, Int), U)] =
-      rightRDD.flatMap(kv => {
-        val (region, y) = kv
-        val lo = bins.value.getStartBin(region)
-        val hi = bins.value.getEndBin(region)
-        (lo to hi).map(i => ((region, i), y))
-      })
+  def compute(): RDD[(RT, RU)] = {
+    leftRdd.zipPartitions(rightRdd, preservesPartitioning = true)(makeIterator)
```

Comment:
That's my point: isn't the `ManualRegionPartitioner` invalid after the `zipPartitions`, since the "key" now has type `RT` and the partitioner expects type `(_, Int)`?